GH-31303: [Python] Remove the legacy ParquetDataset custom python-based implementation

AlenkaF opened the pull request:
### Rationale for this change

Legacy ParquetDataset has been deprecated for a while now, see https://github.com/apache/arrow/issues/31529. This PR is removing the legacy implementation from the code.

### What changes are included in this PR?

The PR is removing:
- `ParquetDatasetPiece `
-  `ParquetManifest`
-  `_ParquetDatasetMetadata `
-  `ParquetDataset`

The PR is renaming `_ParquetDatasetV2` to `ParquetDataset` which was removed. It is also updating the docstrings.

The PR is updating:
- `read_table`
-  `write_to_dataset`

The PR is updating all the tests to not use `use_legacy_dataset` keyword or legacy parametrisation.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Deprecated code is removed.
* Closes: #31303

jorisvandenbossche commented on python/pyarrow/tests/parquet/test_basic.py:
```diff
@@ -63,21 +62,20 @@ def test_parquet_invalid_version(tempdir):
 
-@parametrize_legacy_dataset
-def test_set_data_page_size(use_legacy_dataset):
+@pytest.mark.dataset
```
I know this mark was in practice used through `@parametrize_legacy_dataset`, but we might not need to add it for all cases. Here, this `_check_roundtrip` is using `pq.read_table` under the hood. But for many cases (reading simple files), we have a fallback in `read_table` to use ParquetFile in case the dataset API is not available.

So I think we only need to mark those tests with `@pytest.mark.dataset` if it is actually testing a feature that is only supported in the dataset-based code path (eg reading multiple files instead of a single file, or when specifying a filter or partitioning)

jorisvandenbossche reviewed (commented):
(already posting whatever I have right now)


The `PartitionSet` and `ParquetPartitions` classes can also be removed?

There a few helper methods like `_get_filesystem_and_path` and `_mkdir_if_not_exists` that are no longer used and can be removed as well

AlenkaF commented on python/pyarrow/parquet/core.py:
```diff
@@ -2961,129 +1949,77 @@ def partitioning(self):
+def read_table(source, *, columns=None, use_threads=True,
                schema=None, use_pandas_metadata=False, read_dictionary=None,
                memory_map=False, buffer_size=0, partitioning="hive",
-               filesystem=None, filters=None, use_legacy_dataset=False,
```
Specifying `use_legacy_dataset=False` doesn't break in the current code due to **kwargs kept in `ParquetDataset` from `_ParquetDatasetV2`. I would personally keep it as is as it gets ignored currently and would add a deprecation warning and a test for it.

AlenkaF replied:
Oh sorry, I meant to add this response to the first comment with the same subject: https://github.com/apache/arrow/pull/39112/files#r1418592766

AlenkaF commented:
It is very hard to review this PR due to the way the diff is presented in GitHub. I tried to summarise the main changes in the description of the PR, hope it helps a bit.

@jorisvandenbossche after the last review I have updated the marks in the tests, added `use_legacy_dataset=None` to `ParquetDataset` class, `read_table` and `arite_to_dataset`.

I have also removed `** kwargs` from `ParquetDataset`, previously `_ParquetDatasetV2`, which meant I had to remove the code connected to `metadata`, `split_row_groups` and `validate_schema` ([raising an error](https://github.com/apache/arrow/blob/92e56ba8906f40996cc81bc09fca10c4d53b32fa/python/pyarrow/parquet/core.py#L2441-L2448)) and so I added a note in the docstrings https://github.com/apache/arrow/pull/39112/commits/b6799cfbe68a3289e00378739bc5291ff81d8855. Not sure if that is well done though.

AlenkaF commented:
@github-actions crossbow submit -g python

jorisvandenbossche commented on python/pyarrow/tests/parquet/test_basic.py:
```diff
@@ -752,17 +682,15 @@ def test_fastparquet_cross_compatibility(tempdir):
     lambda: pa.array(["", None] * 10),
     lambda: pa.array(["", None] * 10).dictionary_encode(),
 ])
-@pytest.mark.parametrize('use_dictionary', [False, True])
```
Is there a reason this parametrization was also removed?

jorisvandenbossche replied:
Ah, because it was just not used in practice

jorisvandenbossche commented on python/pyarrow/tests/parquet/test_compliant_nested_type.py:
```diff
@@ -83,21 +79,18 @@ def test_write_compliant_nested_type_enable(tempdir,
 
 @pytest.mark.pandas
-@parametrize_legacy_dataset
+@pytest.mark.dataset
```
Is dataset mark needed here?

jorisvandenbossche commented on python/pyarrow/tests/parquet/test_datetime.py:
```diff
@@ -48,8 +47,8 @@
 @pytest.mark.pandas
-@parametrize_legacy_dataset
-def test_pandas_parquet_datetime_tz(use_legacy_dataset):
+@pytest.mark.dataset
```
Same here and the one below about the dataset marker

jorisvandenbossche commented on python/pyarrow/tests/test_dataset.py:
```diff
@@ -1388,16 +1380,13 @@ def _create_dataset_all_types(tempdir, chunk_size=None):
     path = str(tempdir / "test_parquet_dataset_all_types")
 
     # write_to_dataset currently requires pandas
```
This comment is no longer correct I think? (so that might also mean that all tests that use this helper no longer need the pandas mark)

AlenkaF replied:
Oh yes, it was needed for `partition_cols` in the legacy path: https://github.com/apache/arrow/blob/87865b5a85c722ef7578aed4300e9d0b219c909c/python/pyarrow/parquet/core.py#L3516

Will remove.

jorisvandenbossche commented on python/pyarrow/tests/parquet/test_pandas.py:
```diff
@@ -150,52 +144,22 @@ def test_pandas_parquet_2_0_roundtrip_read_pandas_no_index_written(
     tm.assert_frame_equal(df, df_read)
 
 
-# TODO(dataset) duplicate column selection actually gives duplicate columns now
```
Do you know if this is actually tested elsewhere? (that you get duplicate columns)

AlenkaF replied:
I think there is only one similar test using datasets
https://github.com/apache/arrow/blob/3c66491846a24f17014b31a22fafdda0229f881a/python/pyarrow/tests/test_dataset.py#L3472-L3473
which is same at the end but will still add another for `pq.read_table` which is now using only the new dataset API. Not sure if I have to roundtrip with pandas though?

AlenkaF replied:
https://github.com/apache/arrow/pull/39112/commits/5b1f6eda7b8696989ff4487171770677912c58fb

jorisvandenbossche commented on python/pyarrow/tests/parquet/test_pandas.py:
```diff
@@ -497,15 +447,14 @@ def test_categorical_index_survives_roundtrip(use_legacy_dataset):
 @pytest.mark.pandas
-@parametrize_legacy_dataset
-def test_categorical_order_survives_roundtrip(use_legacy_dataset):
+@pytest.mark.dataset
```
I suppose that many of those tests also don't need the dataset mark

AlenkaF replied:
I removed all dataset marks, looks like the tests are using `pq.write_table`.

jorisvandenbossche commented on python/pyarrow/tests/parquet/test_dataset.py:
```diff
@@ -938,46 +671,26 @@ def test_read_multiple_files(tempdir, use_legacy_dataset):
     assert result.equals(expected)
 
-    # Read with provided metadata
-    # TODO(dataset) specifying metadata not yet supported
```
We actually do support that through a different dataset API (`pyarrow.dataset.parquet_dataset`), but that is not wired up in the parquet API. Given that we never got any complaints about this, I suppose fine to just remove (and it should already be tested in the dataset tests, so should be fine to remove the tests here as well)

jorisvandenbossche commented on python/pyarrow/tests/parquet/test_dataset.py:
```diff
@@ -986,31 +699,24 @@ def read_multiple_files(paths, columns=None, use_threads=True, **kwargs):
-    if not use_legacy_dataset:
-        # TODO(dataset) Dataset API skips bad files
-        return
+    # TODO(dataset) Dataset API skips bad files
```
Do we know if there is an issue about this?

AlenkaF replied:
I haven't found one. I am trying to figure out what should be tested and will update it so that skipping bad files in Dataset API is included in it.

AlenkaF replied:
Can also open an issue and resolve this TODO in a separate PR.

jorisvandenbossche replied:
Yes, I think that's good, let's further figure this out in a follow-up issue

AlenkaF replied:
Opened: https://github.com/apache/arrow/issues/39338

jorisvandenbossche commented on python/pyarrow/tests/parquet/test_dataset.py:
```diff
@@ -1123,18 +819,14 @@ def _make_example_multifile_dataset(base_path, nfiles=10, file_nrows=5):
-        assert set(paths) == set(dataset._dataset.files)
+def _assert_dataset_paths(dataset, paths):
+    paths = [str(path.as_posix()) for path in paths]
+    assert set(paths) == set(dataset._dataset.files)
```
```suggestion
    assert set(paths) == set(dataset.files)
```

(this is exposed nowadays, so we can use the public attribute here in the tests as well)

jorisvandenbossche commented on python/pyarrow/tests/parquet/test_dataset.py:
```diff
@@ -1230,29 +915,22 @@ def test_ignore_custom_prefixes(tempdir, use_legacy_dataset):
-def test_empty_directory(tempdir, use_legacy_dataset):
-    # ARROW-5310 - reading empty directory
-    # fails with legacy implementation
+def test_empty_directory(tempdir):
```
```suggestion
def test_empty_directory(tempdir):
    # ARROW-5310 - reading empty directory
```
Maybe keep the issue reference (that's still relevant for the dataset case, as that was the reason this test was introduced)

jorisvandenbossche reviewed (commented):
Some more comments on the tests!

jorisvandenbossche commented:
I think the minimal builds are not in the "python" group for crossbow (maybe something we should fix), so I think we have to trigger them separately

jorisvandenbossche commented:
@github-actions crossbow submit example-python-minimal-*

jorisvandenbossche commented on python/pyarrow/parquet/core.py:
```diff
@@ -1714,11 +1185,13 @@ def _open_dataset_file(dataset, path, meta=None):
 Encapsulates details of reading a complete Parquet dataset possibly
 consisting of multiple files and partitions in subdirectories.
 
+Keywords `metadata`, `split_row_groups=True` and `validate_schema=False`
+are not yet supported.
```
We could maybe also leave this out? The wording now suggests that we are planning to support them, and I am not sure if that is the case?

jorisvandenbossche commented:
@github-actions crossbow submit -g python

jorisvandenbossche commented:
The HDFS failures seem to be related: https://github.com/ursacomputing/crossbow/actions/runs/7288206604/job/19860362598#step:6:9473

AlenkaF commented:
Yeah, missed some metadata there. Thanks, will correct!

AlenkaF commented:
@jorisvandenbossche if I am understanding correctly `read_parquet` from the failed build is part of a legacy `Filesystem` and will be removed? So I can for time being add a `ValueError` if `metadata` is specified or I can just remove it from the signature?

AlenkaF commented:
@github-actions crossbow submit -g python-*-hdfs

jorisvandenbossche commented:
I am not sure if raising for the `metadata` will be sufficient? My understanding is that those tests (for the legacy HDFS filesystem) are using hdfs.read_parquet -> pq.read_table by passing a legacy HDFS filesystem object, and with removing the legacy parquet code, we no longer support the legacy filesystems (and now as a next step, we can actually also remove those legacy filesystems). 
So maybe we can just "xfail" those tests for now? And then when removing the HDFS legacy filesystem, we can see if we just remove those tests, or if we rewrite them using the new hdfs filesystem.

AlenkaF commented:
Still need to look into the failures with `Unrecognized filesystem` error.

AlenkaF commented:
Ah, I guess the issue is the same it only fails in `_ensure_filesystem` and not in `read_parquet`. Will add xfail marks there also.

AlenkaF commented:
@github-actions crossbow submit *hdfs*

jorisvandenbossche commented:
Thanks a lot @AlenkaF for the work here!

jorisvandenbossche commented:
The `wide-dataframe` case seems a genuine perf regression (and not a flaky outlier as the other listed cases). That might mean that for wide dataframes, the new code path is slower compared to the legacy dataset reader (since with this commit, also when specifying `use_legacy_dataset=True`, the new code path will be used). 
That seems to match with the timing in the `use_legacy_dataset=False` case of the wide-dataframe benchmark, as now both benchmarks more or less show the same timing.

However, I can't reproduce this locally with pyarrow 14.0 (where the legacy reader still exists):

```
import numpy as np
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq

dataframe = pd.DataFrame(np.random.rand(100, 10000))
table = pa.Table.from_pandas(dataframe)
pq.write_table(table, "test_wide_dataframe.parquet")
```

```
In [7]: %timeit -r 50 pq.read_table("test_wide_dataframe.parquet", use_legacy_dataset=True)
392 ms ± 4.67 ms per loop (mean ± std. dev. of 50 runs, 1 loop each)

In [8]: %timeit -r 50 pq.read_table("test_wide_dataframe.parquet", use_legacy_dataset=False)
350 ms ± 11.5 ms per loop (mean ± std. dev. of 50 runs, 1 loop each)
```
